dali/python/nvidia/dali/ops.py

```diff
@@ -601,29 +601,103 @@ def _choose_device(inputs):
-        if dev == "gpu":
-            dev_inputs = list(input.gpu() for input in inputs)
+def _is_integer_like(input):
+    if type(input) == int:
+        return True
+    if isinstance(input, Constant):
+        if input.dtype in [DALIDataType.INT8, DALIDataType.INT16, DALIDataType.INT32,
+                DALIDataType.INT64, DALIDataType.UINT8, DALIDataType.UINT16, DALIDataType.UINT32,
+                DALIDataType.UINT64]:
+            return True
+    return False
+
+def _is_real_like(input):
+    if type(input) == float:
+        return True
+    if isinstance(input, Constant):
+        if input.dtype in [DALIDataType.FLOAT16, DALIDataType.FLOAT, DALIDataType.FLOAT64]:
+            return True
+    return False
+
+def _to_type_desc(input):
+    if  type(input) == int:
+        return "int32"
+    if type(input) == float:
+        return "float32" # TODO(klecki): current DALI limitation
+    if isinstance(input, Constant):
+        dtype_to_desc = {
+            DALIDataType.INT8: "int8",
+            DALIDataType.INT16: "int16",
+            DALIDataType.INT32: "int32",
+            DALIDataType.INT64: "int64",
+            DALIDataType.UINT8: "uint8",
+            DALIDataType.UINT16: "uint16",
+            DALIDataType.UINT32: "uint32",
+            DALIDataType.UINT64: "uint64",
+            DALIDataType.FLOAT16: "float16",
+            DALIDataType.FLOAT: "float32",
+            DALIDataType.FLOAT64: "float64",
+        }
+        return dtype_to_desc[input.dtype]
+    raise TypeError(("Argument to arithmetic operation not supported. Got {}, expected "
+                "a return value from other DALI Operator or a constant value of type 'int', "
+                "'float' or 'nvidia.dali.Constant'.").format(str(type(input))))
+
+
+# Group inputs into _EdgeReferences, integer constants and real constants
+def _group_inputs(inputs):
+    categories_idxs = []
+    edges = []
+    integers = []
+    reals = []
+    for input in inputs:
+        if isinstance(input, _EdgeReference):
+            categories_idxs.append(("edge", len(edges)))
+            edges.append(input)
+        elif _is_integer_like(input):
+            categories_idxs.append(("integer", len(integers)))
+            integers.append(input)
+        elif _is_real_like(input):
+            categories_idxs.append(("real", len(reals)))
+            reals.append(input)
         else:
-            dev_inputs = inputs
-        # Call it imediatelly
-        return op(*dev_inputs)
-
+            raise TypeError(("Argument to arithmetic operation not supported. Got {}, expected "
+                    "a return value from other DALI Operator  or a constant value of type 'int', "
+                    "'float' or 'nvidia.dali.Constant'.").format(str(type(input))))
+    if len(integers) == 0:
+        integers = None
+    if len(reals) == 0:
+        reals = None
+    return (categories_idxs, edges, integers, reals)
+
+def _generate_input_desc(categories_idx, integers, reals):
```

Comment:
I would put some comment what is happening here - like building an expression with some syntax. There is one description in the native code, but probably the python side is the first that user will check looking for details.

Reply:
The user should not be looking for it here, as this is an internal function and he is not to create such description by hand.
I added a comment pointing to ArithmeticGenericOp.

Reply:
I know, but still someone may wan to check internals what is not that uncommon for python code.